File: HWs/RLS-Activity/scripts/r/ejercicio1.R
# %%
library(readr)
library(ggplot2)
library(dplyr)
library(broom)

df <- read_csv("../../data/datos_fuerza_levantamiento.csv")
head(df)

# %%
cat("Generando gráfico de dispersión...\n")

output_dir <- "../../plots/r/ejercicio1"
if (!dir.exists(output_dir)) {
  dir.create(output_dir, recursive = TRUE)
}

p <- ggplot(df, aes(x = Fuerza_del_brazo_x, y = Levantamiento_dinamico_y)) +
  geom_point() +
  labs(x = "Fuerza del brazo, x", 
       y = "Levantamiento dinámico, y",
       title = "Dispersión: y vs. x") +
  theme_minimal() +
  theme(panel.grid.minor = element_line(linetype = "dashed"))

ggsave(file.path(output_dir, "scatter_plot.png"), p, dpi = 300, width = 6, height = 4)
print(p)

# %%
cat("Generando estadísticas descriptivas...\n")
desc <- df %>%
  select(Fuerza_del_brazo_x, Levantamiento_dinamico_y) %>%
  summary()
print(desc)

# %%
cat("Calculando coeficiente de correlación de Pearson...\n")
x <- df$Fuerza_del_brazo_x
y <- df$Levantamiento_dinamico_y

cor_result <- cor.test(x, y, method = "pearson")
r <- cor_result$estimate
pval <- cor_result$p.value

cat("r (Pearson) =", round(r, 4), "\n")
cat("p-valor =", round(pval, 6), "\n")

cat("Realizando prueba de hipótesis...\n")
alpha <- 0.05
if (pval < alpha) {
  cat("Conclusión: se rechaza H0; evidencia de relación lineal (α=0.05).\n")
} else {
  cat("Conclusión: no se rechaza H0; no hay evidencia suficiente de relación lineal (α=0.05).\n")
}

# %%
model <- lm(Levantamiento_dinamico_y ~ Fuerza_del_brazo_x, data = df)
cat("Parámetros estimados:\n")
print(coef(model))

cat("\nResumen del modelo:\n")
print(summary(model))

# %%
cat("Estimando modelo de regresión lineal simple por MCO...\n")
x0 <- 30.0

new_data <- data.frame(Fuerza_del_brazo_x = x0)
pred <- predict(model, new_data, interval = "confidence", level = 0.95)
pred_obs <- predict(model, new_data, interval = "prediction", level = 0.95)

cat("Estimación puntual E[y|x=", x0, "] =", round(pred[1], 4), "\n")
cat("Intervalo de confianza al 95% para la media condicional:\n")
print(pred)

cat("\nIntervalo de predicción al 95% (para una observación nueva):\n")
print(pred_obs)

# %%
resid <- residuals(model)
cat("Generando gráfico de residuales...\n")

p_resid <- ggplot(data.frame(x = x, resid = resid), aes(x = x, y = resid)) +
  geom_point() +
  geom_hline(yintercept = 0, linetype = "dashed") +
  labs(x = "Fuerza del brazo, x", 
       y = "Residuales",
       title = "Residuales vs. x") +
  theme_minimal() +
  theme(panel.grid.minor = element_line(linetype = "dashed"))

ggsave(file.path(output_dir, "residuals_plot.png"), p_resid, dpi = 300, width = 6, height = 4)
print(p_resid)

cat("Media de residuales (debe ser ~0):", round(mean(resid), 6), "\n")
cat("Desviación estándar de residuales:", round(sd(resid), 6), "\n")
corr_rx <- cor(x, resid)
cat("Correlación(x, residuales) =", round(corr_rx, 6), "(debe ser cercana a 0 en MCO con intercepto)\n")

# %%

File: HWs/RLS-Activity/scripts/matlab/ejercicio1.m
%% 

df = readtable('../../data/datos_fuerza_levantamiento.csv');
head(df)

%%
fprintf('Generando gráfico de dispersión...\n');
figure;
scatter(df.Fuerza_del_brazo_x, df.Levantamiento_dinamico_y);
xlabel('Fuerza del brazo, x');
ylabel('Levantamiento dinámico, y');
title('Dispersión: y vs. x');
grid on;
grid minor;


output_dir = '../../plots/matlab/ejercicio1';
if ~exist(output_dir, 'dir')
    mkdir(output_dir);
end
saveas(gcf, fullfile(output_dir, 'scatter_plot.png'), 'png');

%%
fprintf('Generando estadísticas descriptivas...\n');
desc = summary(df(:, {'Fuerza_del_brazo_x', 'Levantamiento_dinamico_y'}));
disp(desc);

%%
fprintf('Calculando coeficiente de correlación de Pearson...\n');
x = df.Fuerza_del_brazo_x;
y = df.Levantamiento_dinamico_y;

[r, pval] = corr(x, y);
fprintf('r (Pearson) = %.4f\n', r);
fprintf('p-valor = %.6f\n', pval);

fprintf('Realizando prueba de hipótesis...\n');
alpha = 0.05;
if pval < alpha
    fprintf('Conclusión: se rechaza H0; evidencia de relación lineal (α=0.05).\n');
else
    fprintf('Conclusión: no se rechaza H0; no hay evidencia suficiente de relación lineal (α=0.05).\n');
end

%%

X = [ones(length(x), 1), x];
[beta, ~, ~, ~, stats] = regress(y, X);
fprintf('Parámetros estimados:\n');
fprintf('Intercepto: %.4f\n', beta(1));
fprintf('Pendiente: %.4f\n', beta(2));

fprintf('\nResumen del modelo:\n');
fprintf('R^2: %.4f\n', stats(1));
fprintf('F-statistic: %.4f\n', stats(2));
fprintf('p-value: %.6f\n', stats(3));
fprintf('Error estándar: %.4f\n', sqrt(stats(4)));

%%
fprintf('Estimando modelo de regresión lineal simple por MCO...\n');
x0 = 30.0;


y_pred = beta(1) + beta(2) * x0;


n = length(x);
x_bar = mean(x);
Sxx = sum((x - x_bar).^2);
MSE = stats(4);


se_mean = sqrt(MSE * (1/n + (x0 - x_bar)^2 / Sxx));
t_crit = tinv(0.975, n-2);


ci_lower = y_pred - t_crit * se_mean;
ci_upper = y_pred + t_crit * se_mean;


se_pred = sqrt(MSE * (1 + 1/n + (x0 - x_bar)^2 / Sxx));
pi_lower = y_pred - t_crit * se_pred;
pi_upper = y_pred + t_crit * se_pred;

fprintf('Estimación puntual E[y|x=%.1f] = %.4f\n', x0, y_pred);
fprintf('Intervalo de confianza al 95%%:\n');
fprintf('[%.4f, %.4f]\n', ci_lower, ci_upper);

fprintf('\nIntervalo de predicción al 95%% (para una observación nueva):\n');
fprintf('[%.4f, %.4f]\n', pi_lower, pi_upper);

%%
resid = y - (beta(1) + beta(2) * x);
fprintf('Generando gráfico de residuales...\n');
figure;
scatter(x, resid);
hold on;
yline(0, '--');
xlabel('Fuerza del brazo, x');
ylabel('Residuales');
title('Residuales vs. x');
grid on;
grid minor;

saveas(gcf, '../../plots/matlab/ejercicio1/residuals_plot.png', 'png');

fprintf('Media de residuales (debe ser ~0): %.6f\n', mean(resid));
fprintf('Desviación estándar de residuales: %.6f\n', std(resid, 1));
corr_rx = corr(x, resid);
fprintf('Correlación(x, residuales) = %.6f (debe ser cercana a 0 en MCO con intercepto)\n', corr_rx);

%%
